File: backend/license_verify.py
"""Offline Ed25519 license verification (same format as Electron).

Deliberately does not check any machine/device binding: the signed payload
carries no machine_id (unknown at signing time — the client just pastes a key
with no prior ID exchange). Device binding is enforced server-side, once, at
first-activation time by cloud-node (routes/licenses.js); this module only
re-checks the cheap, fully-offline parts (signature, product, tier, seats) on
every local entitlement read.
"""

from __future__ import annotations

import base64
import json
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from entitlement_constants import EMBEDDED_LICENSE_PUBLIC_KEY_HEX, LICENSE_PREFIX, PRODUCT_SLUG


def _canonical_payload(obj: dict[str, Any]) -> str:
    keys = sorted(obj.keys())
    ordered = {k: obj[k] for k in keys}
    return json.dumps(ordered, separators=(",", ":"), sort_keys=True)


def _b64url_decode(s: str) -> bytes:
    pad = "=" * ((4 - len(s) % 4) % 4)
    return base64.urlsafe_b64decode(s + pad)


def verify_license_key(license_key: str) -> tuple[bool, str | None, dict | None]:
    """Returns (ok, reason_code, payload_or_none)."""
    trimmed = (license_key or "").strip()
    if not trimmed:
        return False, "empty", None
    parts = trimmed.split(".")
    if len(parts) != 3 or parts[0] != LICENSE_PREFIX:
        return False, "format", None
    try:
        raw = _b64url_decode(parts[1]).decode("utf-8")
        payload = json.loads(raw)
    except Exception:
        return False, "payload", None
    if not isinstance(payload, dict):
        return False, "payload", None
    if payload.get("product") != PRODUCT_SLUG:
        return False, "product", None
    if payload.get("tier") != "full":
        return False, "tier", None
    if not isinstance(payload.get("license_id"), str) or not payload.get("license_id"):
        return False, "license_id", None
    max_seats = payload.get("max_seats")
    if not isinstance(max_seats, int) or isinstance(max_seats, bool) or max_seats < 1:
        return False, "max_seats", None
    message = _canonical_payload(payload).encode("utf-8")
    try:
        sig = _b64url_decode(parts[2])
    except Exception:
        return False, "sig_format", None
    if len(sig) != 64:
        return False, "sig_len", None
    pub_bytes = bytes.fromhex(EMBEDDED_LICENSE_PUBLIC_KEY_HEX)
    pub = Ed25519PublicKey.from_public_bytes(pub_bytes)
    try:
        pub.verify(sig, message)
    except InvalidSignature:
        return False, "sig_verify", None
    return True, None, payload

File: cloud-node/lib/licenseVerify.js
/**
 * Server-side counterpart to `electron/entitlement/verify.js` /
 * `backend/license_verify.py`. Unlike those, this never checks a `machine_id`
 * against "the current device" — the server has no such concept. Device
 * binding is enforced by `licenseActivations.js` instead, keyed by
 * `payload.license_id`.
 */

const ed = require("@noble/ed25519");
const { LICENSE_PREFIX, PRODUCT_SLUG } = require("./licenseConstants");
// Not destructured: tests reassign `licenseConstants.EMBEDDED_LICENSE_PUBLIC_KEY_HEX`
// to verify against a throwaway keypair instead of the real embedded one.
const licenseConstants = require("./licenseConstants");

/** Same canonical form as the Electron/Python signers. */
function canonicalLicensePayload(obj) {
  const ordered = {};
  for (const k of Object.keys(obj).sort()) {
    ordered[k] = obj[k];
  }
  return JSON.stringify(ordered);
}

/**
 * @param {string} licenseKey
 * @returns {Promise<{ ok: boolean, reason?: string, payload?: { license_id: string, max_seats: number, product: string, tier: string, iat: number } }>}
 */
async function verifyLicenseKey(licenseKey) {
  const trimmed = typeof licenseKey === "string" ? licenseKey.trim() : "";
  if (!trimmed) {
    return { ok: false, reason: "empty" };
  }
  const parts = trimmed.split(".");
  if (parts.length !== 3 || parts[0] !== LICENSE_PREFIX) {
    return { ok: false, reason: "format" };
  }
  let payload;
  try {
    payload = JSON.parse(Buffer.from(parts[1], "base64url").toString("utf8"));
  } catch {
    return { ok: false, reason: "payload" };
  }
  if (!payload || typeof payload !== "object") {
    return { ok: false, reason: "payload" };
  }
  if (payload.product !== PRODUCT_SLUG) {
    return { ok: false, reason: "product" };
  }
  if (payload.tier !== "full") {
    return { ok: false, reason: "tier" };
  }
  if (typeof payload.license_id !== "string" || !payload.license_id) {
    return { ok: false, reason: "license_id" };
  }
  const maxSeats = Number(payload.max_seats);
  if (!Number.isInteger(maxSeats) || maxSeats < 1) {
    return { ok: false, reason: "max_seats" };
  }
  let sig;
  try {
    sig = Buffer.from(parts[2], "base64url");
  } catch {
    return { ok: false, reason: "sig_format" };
  }
  if (sig.length !== 64) {
    return { ok: false, reason: "sig_len" };
  }
  const message = new TextEncoder().encode(canonicalLicensePayload(payload));
  const pub = Uint8Array.from(Buffer.from(licenseConstants.EMBEDDED_LICENSE_PUBLIC_KEY_HEX, "hex"));
  const valid = await ed.verifyAsync(Uint8Array.from(sig), message, pub);
  if (!valid) {
    return { ok: false, reason: "sig_verify" };
  }
  return { ok: true, payload };
}

module.exports = { verifyLicenseKey, canonicalLicensePayload };
